File: 4.2-procesamiento-datos/prueba-2/version-2/python/main.py
import js  # type: ignore
from pyscript import display, PyWorker, document
import json
import time
import asyncio

workers = []
workers_ready = False


async def initialize_worker():
    global workers, workers_ready
    num_workers = int(js.document.querySelector(
        "#parallel-workers-pyscript").value)

    # Si ya hay trabajadores, eliminarlos antes de crear nuevos
    if len(workers) > num_workers:
        workers = workers[:num_workers]  # Mantener solo los necesarios
    elif len(workers) < num_workers:
        # Crear los trabajadores faltantes
        additional_workers = num_workers - len(workers)
        for _ in range(additional_workers):
            worker = PyWorker(
                "./python/worker.py",
                type="pyodide",
                config="./json/pyscript-worker.json"
            )
            await worker.ready
            workers.append(worker)

    workers_ready = True  # Confirmar que los trabajadores están listos

asyncio.create_task(initialize_worker())


async def run_py_benchmark(event):
    global workers
    await initialize_worker()
    try:
        num_executions = int(js.document.querySelector(
            "#num-executions-pyscript").value)
        num_workers = int(js.document.querySelector(
            "#parallel-workers-pyscript").value)

        # Asegurar que tenemos la cantidad correcta de trabajadores antes de ejecutar
        await initialize_worker()

        js.clearCell("pyscript")
        start_time = time.perf_counter()

        # Distribuir las tareas entre los trabajadores disponibles
        tasks = []
        for i in range(num_executions):
            # Evita acceder a un índice fuera de rango
            worker = workers[i % num_workers]
            tasks.append(worker.sync.do_statistical_analysis(10_000_000))

        # Ejecutar en paralelo
        results_list = await asyncio.gather(*tasks)
        print("📥 Resultados recibidos en run_py_benchmark:", results_list)

        accumulated = {
            'create': {'time': 0.0, 'memory': 0.0},
            'sum': {'time': 0.0, 'memory': 0.0},
            'mean': {'time': 0.0, 'memory': 0.0},
            'std': {'time': 0.0, 'memory': 0.0},
            'total_per_execution': 0.0,
        }

        # Procesar resultados
        for result in results_list:
            data = json.loads(result)
            for op in ['create', 'sum', 'mean', 'std']:
                accumulated[op]['time'] += data[op]['time']
                accumulated[op]['memory'] += data[op]['memory']
            accumulated['total_per_execution'] += data['total']['time']

        # Promediar resultados
        results = {op: {
            'time': accumulated[op]['time'] / num_executions,
            'memory': accumulated[op]['memory'] / num_executions
        } for op in ['create', 'sum', 'mean', 'std']}

        results['total'] = {
            'average_per_execution': accumulated['total_per_execution'] / num_executions,
            'total_time': (time.perf_counter() - start_time) * 1000
        }

        # Actualizar UI
        update_ui(results)

    except Exception as e:
        display(f"Error: {str(e)}", target="pyscript-output")


def update_ui(results):
    for operation in ['create', 'sum', 'mean', 'std']:
        display(
            f"{results[operation]['time']:.2f} | {results[operation]['memory']: .2f}",
            target=f"pyscript-{operation}"
        )

    display(
        f"{results['total']['average_per_execution']:.2f}",
        target="pyscript-output"
    )
    display(
        f"{results['total']['total_time']:.2f}",
        target="pyscript-exact"
    )


def js_run_py_benchmark(event):
    asyncio.ensure_future(run_py_benchmark(None))

File: 4.2-procesamiento-datos/prueba-2/version-2/javascript/main.js
function runJSBenchmark() {
    const numExecutions = parseInt(document.getElementById("num-executions-javascript").value) || 1;
    clearCell("javascript");
    let simulationStartTime = performance.now();

    const promises = [];

    for (let i = 0; i < numExecutions; i++) {
        promises.push(new Promise((resolve, reject) => {
            const worker = new Worker("./javascript/worker.js");

            worker.onmessage = function (event) {
                if (event.data.error) {
                    console.error("Error in JS Worker:", event.data.error, event.data.stack);
                    reject(event.data.error);
                } else {
                    resolve(event.data.results);
                }
                worker.terminate();
            };

            worker.onerror = function (error) {
                console.error("JS Worker Error:", error);
                reject(error);
                worker.terminate();
            };

            const dataSize = 10_000_000;
            worker.postMessage({ size: dataSize });
        }));
    }

    Promise.all(promises)
        .then((resultsArray) => {
            let totalResults = {
                create: { time: 0, memory: 0 },
                sum: { time: 0, memory: 0 },
                mean: { time: 0, memory: 0 },
                std: { time: 0, memory: 0 },
                total: { time: 0, memory: 0, total_time_exact: 0 }
            };

            resultsArray.forEach(results => {
                for (let op in totalResults) {
                    if (results[op]) {
                        totalResults[op].time += results[op].time;
                        totalResults[op].memory += results[op].memory;
                    }
                }
            });

            for (let op in totalResults) {
                totalResults[op].time /= numExecutions;
                totalResults[op].memory /= numExecutions;
            }

            totalResults.total.total_time_exact = performance.now() - simulationStartTime;

            displayResults(totalResults);
        })
        .catch((error) => {
            console.error("Worker error:", error);
            displayJSText("javascript-output", `Worker Error: ${error}`);
        });
}

function displayResults(results) {
    displayJSText("javascript-create", `${results.create.time.toFixed(2)} ms | ${results.create.memory.toFixed(2)} MB`);
    displayJSText("javascript-sum", `${results.sum.time.toFixed(2)} ms | ${results.sum.memory.toFixed(2)} MB`);
    displayJSText("javascript-mean", `${results.mean.time.toFixed(2)} ms | ${results.mean.memory.toFixed(2)} MB`);
    displayJSText("javascript-std", `${results.std.time.toFixed(2)} ms | ${results.std.memory.toFixed(2)} MB`);
    displayJSText("javascript-output",
        `TOTAL (Promedio): ${results.total.time.toFixed(2)} ms | Peak: ${results.total.memory.toFixed(2)} MB`
    );
    displayJSText("javascript-exact",
        `TOTAL Exacto: ${results.total.total_time_exact.toFixed(2)} ms`
    );
}

function clearCell(elementId) {
    const operations = ['create', 'sum', 'mean', 'std', 'output', 'exact'];
    operations.forEach(op => {
        const element = document.getElementById(`${elementId}-${op}`);
        if (element) {
            element.innerHTML = "";
        }
    });
}

function displayJSText(elementId, text) {
    const element = document.getElementById(elementId);
    if (element) {
        element.textContent += text;
    } else {
        console.warn(`Element with ID ${elementId} not found.`);
    }
}
